java-tests/validating/src/test/kotlin/io/spine/test/options/setonce/SetOnceErrorMessageITest.kt

```diff
@@ -0,0 +1,173 @@
+
+@DisplayName("`(set_once)` constraint should")
+internal class SetOnceErrorMessageITest {
+
+    @MethodSource("allFieldTypesWithTwoDistinctValues")
+    @ParameterizedTest(name = "show the default error message for `{0}` field")
+    fun <T : Any> defaultErrorMessage(fieldName: String, value1: T, value2: T) =
+        assertDefaultMessage(fieldName, value1, value2)
+
+    @MethodSource("allFieldTypesWithTwoDistinctValues")
+    @ParameterizedTest(name = "show the custom error message for `{0}` field")
+    fun <T : Any> customErrorMessage(fieldName: String, value1: T, value2: T) =
+        assertCustomMessage(fieldName, value1, value2)
+
+    private companion object {
+
+        @JvmStatic
+        fun allFieldTypesWithTwoDistinctValues() = listOf(
+            arguments(named("message", "name"), JACK, DONALD),
+            arguments(named("string", "id"), STUDENT1, STUDENT2),
+            arguments(named("double", "height"), SHORT_HEIGHT, TALL_HEIGHT),
+            arguments(named("float", "weight"), FIFTY_KG, EIGHTY_KG),
+            arguments(named("int32", "cash_USD"), SIXTEEN, SIXTY),
+            arguments(named("int64", "cash_EUR"), EIGHTEEN, EIGHTY),
+            arguments(named("uint32", "cash_JPY"), SIXTEEN, SIXTY),
+            arguments(named("uint64", "cash_GBP"), EIGHTEEN, EIGHTY),
+            arguments(named("sint32", "cash_AUD"), SIXTEEN, SIXTY),
+            arguments(named("sint64", "cash_CAD"), EIGHTEEN, EIGHTY),
+            arguments(named("fixed32", "cash_CHF"), SIXTEEN, SIXTY),
+            arguments(named("fixed64", "cash_CNY"), EIGHTEEN, EIGHTY),
+            arguments(named("sfixed32", "cash_PLN"), SIXTEEN, SIXTY),
+            arguments(named("sfixed64", "cash_NZD"), EIGHTEEN, EIGHTY),
+            arguments(named("bool", "has_medals"), YES, NO),
+            arguments(named("enum", "signature"), CERF1, CERF2),
+
+            // For some reason, for enums, `Message.Builder.setField()` expects value
+            // descriptors instead of constants or their ordinal numbers.
+            arguments("year_of_study", FIRST_YEAR.valueDescriptor, THIRD_YEAR.valueDescriptor)
+        )
+    }
+}
+
+private fun <T : Any> assertDefaultMessage(fieldName: String, value1: T, value2: T) {
+    val builder = StudentDefaultMessage.newBuilder()
+    val expectedParams = listOf(fieldName, "$value1", "$value2")
+    val expectedFormat = { _: Int -> DEFAULT_MESSAGE_FORMAT }
+    return builder.assertErrorMessage(fieldName, value1, value2, expectedParams, expectedFormat)
+}
+
+private fun <T : Any> assertCustomMessage(fieldName: String, value1: T, value2: T) {
+    val builder = StudentCustomMessage.newBuilder()
+    val expectedParams = listOf("$value1", fieldName, "$value2")
+    val expectedFormat = ::customMessageFormat
+    return builder.assertErrorMessage(fieldName, value1, value2, expectedParams, expectedFormat)
+}
+
+/**
+ * Asserts that this [GeneratedMessage] throws [ValidationException] with
```

Comment:
This is not a `GeneratedMessage`, this is its `Builder`.

BTW, please consider a less specific type for creating an extension function, `GeneratedMessageV3` is way too narrow, and potentially, kills any of fantastic plans to support Protobuf Lite.